opm/common/utility/CSRGraphFromCoordinates_impl.hpp

```diff
@@ -0,0 +1,546 @@
+{
+    // We assume that we're only called *after* sortColumnIndicesPerRow()
+    // whence duplicate elements appear consecutively in [begin, end).
+    //
+    // Note: This is essentially the same as std::unique(begin, end) save
+    // for the return value and the fact that we additionally record the
+    // 'compressedIdx_' mapping.  That mapping enables subsequent, decoupled
+    // accumulation of the 'sa_' contributions.
+
+    while (begin != end) {
+        // Note: Order of ja_ and compressedIdx_ matters here.
+
+        if constexpr (TrackCompressedIdx) {
+            this->compressedIdx_.push_back(this->ja_.size());
+        }
+
+        this->ja_.push_back(*begin);
+
+        auto next_unique =
+            std::find_if(begin, end, [last = this->ja_.back()]
+                         (const auto j) { return j != last; });
+
+        if constexpr (TrackCompressedIdx) {
+            // Number of duplicate elements in [begin, next_unique).
+            const auto ndup = std::distance(begin, next_unique);
+
+            if (ndup > 1) {
+                // Insert ndup - 1 copies of .back() to represent the
+                // duplicate pairs in [begin, next_unique).  We subtract one
+                // to account for .push_back() above representing *begin.
+                this->compressedIdx_.insert(this->compressedIdx_.end(),
+                                            ndup - 1,
+                                            this->compressedIdx_.back());
+            }
+        }
+
+        begin = next_unique;
+    }
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::CSR::
+remapCompressedIndex([[maybe_unused]] Start&& compressedIdx)
+{
+    if constexpr (TrackCompressedIdx) {
+        for (auto& i : compressedIdx) {
+            i = this->compressedIdx_[i];
+        }
+
+        this->compressedIdx_.swap(compressedIdx);
+    }
+}
+
+// =====================================================================
+
+// ---------------------------------------------------------------------
+// Class Opm::utility::CSRGraphFromCoordinates
+// ---------------------------------------------------------------------
+
+template <typename VertexID, bool TrackCompressedIdx>
+void Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::clear()
+{
+    this->uncompressed_.clear();
+    this->csr_.clear();
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+addConnection(const VertexID v1, const VertexID v2)
+{
+    if ((v1 < 0) || (v2 < 0)) {
+        throw std::invalid_argument {
+            "Vertex IDs must be non-negative.  Got (v1,v2) = ("
+            + std::to_string(v1) + ", " + std::to_string(v2)
+            + ')'
+        };
+    }
+
+    if (v1 == v2) {
+        // Ignore self connections.
```

Comment:
This is not 100% obvious, so I think a comment should be added to the addConnection() doc in the header file.